contribute/design/1.x/migrate/spc-to-cspc-migration.md

```diff
@@ -0,0 +1,205 @@
+- Phase 1: Ability to perform migration from SPC to CSPC using
+  a Kubernetes Job.
+- Phase 2: Allow for migration of old CStor volumes to CSI CStor volumes.
+
+## Motivation
+
+### Goals
+
+- Ease the process of migrating SPC to CSPC by automating it via kubernetes job.
+- Enable day 2 operations on CStor Volumes by migrating them to CSI volumes.
+
+## Proposal
+
+### Proposed Implementation
+
+This design proposes the following key changes while migrating from a SPC to CSPC:
+
+  1. The SPC label and finalizer on BDCs will be replaced by CSPC label and finalizer. This is done to avoid webhook validation failure while creating equivalent CSPC.
+
+  2. Equivalent CSPC CR will be created for the migrating SPC.
+
+  3. Equivalent CSPI CRs will be created by operator which will replace the CSP for given SPC. The CSPI will reconcile disabled to avoid double import.
+
+  4. The SPC owner reference on the BDCs will be replaced by equivalent CSPC owner reference.
+
+  5. Sequentially for each CSP
+
+      1. Old CSP deployment will be scaled down and sync will be enabled on equivalent CSPI.
+      2. Old pool will be imported by renaming it from `cstor-cspuid` to `cstor-cspcuid`.
+      3. The CVR with CSP labels and annotations will be replaced by CSPI labels and annotations.
+      4. Finalizers from CSP will be removed to for proper cleanup.
+
+  6. Clean up old SPC and CSP after successfull creation of CSPC and CSPI objects.
+
+For migrating non-csi volumes to csi volumes following changes are proposed:
+
+  1. New StorageClass with CSPC will be created and this will replace the old strageclass in the PVC of CStorVolume.
+
+  2. Set the PV to `Retain` reclaim policy to prevent deletion of OpenEBS resources.
+
+  3. Recreate old PV with csi spec translation.
+
+  4. Recreate old PVC with updated csi spec PV and StorageClass.
+
+  5. Create CVC CR for the volume with the CV details.
+
+  6. Update ownerreferences of CV, service with CVC and CVRs with CV.
+
+### High Level Design
+
+#### Phase 1: Migration of SPC to CSPC
+
+The migration of SPC will be performed via a job which takes SPC name as one of its argument.
+
+The CSPC CR created via job will have `reconcile.openebs.io/disable-dependants` annotation set to  `true`. This will help in disabling reconciliation of on all CSPIs created for the CSPC. The reconciliation is set off on CSPIs to avoid import while the old CSP pods are still running. Once all CSPI are successfully created the annotation will be removed.
+
+Sequentially one CSPI is taken and the corresponding CSP is found using `kubernetes/hostname` label. The CSP deployment is scaled down to avoid multiple pods trying to import the same pool. Next the all the BDC for given CSPI are updated with CSPC information. The CSPI will be patched with the annotation `cstorpoolinstance.openebs.io/oldname`. Then CSPI reconciliation will be enabled which will create the CSPI deployment which will rename and import the pool. 
```

Comment:
> Then CSPI reconciliation will be enabled which will create the CSPI deployment which will rename and import the pool. 

So CSPI deploymets will be created after enabling the reconciliation? I am guessing it should be Then CSPI reconciliation will be enabled then the pool manager will import the pool.

Reply:
No CSPI deployment is created by operator at the time of creation itself. But the CSPI will have reconciliation disabled so import will not happen. Once the reconciliation is enabled then only it will import the pool.